### PR TITLE
SI-5479 deprecate DelayedInit outside of App

### PR DESCRIPTION
**Pull request by adriaanm**
DelayedInit's semantics are way too surprising.

For example, it delays initialization of fields,
so that fields on objects that extend `App`
(which `extends DelayedInit`) are not initialized
until the `main` method is called.

For more details and a proposed alternative,
see https://issues.scala-lang.org/browse/SI-4330?jql=labels%20%3D%20delayedinit%20AND%20resolution%20%3D%20unresolved.

Support for `App` will continue -- we'll special case it.

review by @gkossakowski

### CODE REVIEW

**Comment by som-snytt** [comment]
Should I retract my comment on 

http://stackoverflow.com/a/21885076/1296806

```
Maybe it's the cargo-cultist in me talking, but this strikes me as a nice application of the DelayedInit magic.
```

I also mentioned the downside of deprecation.

**Comment by adriaanm** [comment]
Would SI-4330's postCreate cover that use case? I must admit I'm getting a bit bleary-eyed from deprecating all this dusty code, so I didn't look very closely.

**Comment by som-snytt** [comment]
That makes me want to run over with cucumber and avocado facial mask, but the use case was to wrap init code in breakable; but you could imagine anything that looks like the App use case, namely, I have init code that I want to handle as an ordinary block.

Famously, REPL code translates to members but block semantics is more intuitive. I wonder if REPL templates should be DelayedInit?  I haven't investigated (yet).

Furthermore, it was hard enough when my daughter was singing all the songs from Frozen all day long, over and over, but she heard the Pink duet Give Me a Reason in the car and then found it on youtube so now she's singing that over and over and now it's going through my head.  Does DelayedInit address that use case? She's not supposed to be searching youtube by herself.

**Comment by adriaanm** [comment]
That's one initiation I imagine every parent would like to delay.

File: src/library/scala/DelayedInit.scala [hunk 1]
```diff
@@ -43,6 +43,7 @@ package scala
  *
  *  @author  Martin Odersky
  */
+@deprecated("DelayedInit semantics can be surprising.\n(For details and a proposed alternative, see https://issues.scala-lang.org/browse/SI-4330?jql=labels%20%3D%20delayedinit%20AND%20resolution%20%3D%20unresolved. Support for `App` will continue.)", "2.11.0")
```
Comment (xeno-by) [inline src/library/scala/DelayedInit.scala]: The URL seems a bit longer than necessary.

Comment (gkossakowski) [inline src/library/scala/DelayedInit.scala]: How about using https://issues.scala-lang.org/browse/SI-4330?filter=12418 ?

**Comment by gkossakowski** [comment]
I'd prefer if we used saved search which generates less noisy URL: https://issues.scala-lang.org/browse/SI-4330?filter=12418

Other than that it looks great.

**Comment by adriaanm** [comment]
The problem with saved searches is that they can be deleted accidentally.

**Comment by gkossakowski** [comment]
Every problem in computer science can be solved with another layer of indirection. How about linking to release notes where we would have a paragraph or two motivating for that deprecation and from there link to JIRA tickets?

**Comment by adriaanm** [comment]
Where should we link to? https://github.com/scala/scala/releases/tag/v2.11.0-RC1? 
http://www.scala-lang.org/news/2014/??/??/release-notes-2.11.0-RC1.html?

**Comment by gkossakowski** [comment]
Ah, I forgot that our release notes include the date so we can't predict the URL :-/
I guess linking to github is not a bad option.

**Comment by adriaanm** [comment]
I included a draft of the blurb in the release notes already: https://github.com/scala/make-release-notes/pull/50

**Comment by adriaanm** [comment]
Pushed a new commit.

File: test/files/neg/delayed-init-ref.check [hunk 2]
```diff
@@ -4,9 +4,13 @@ delayed-init-ref.scala:17: warning: Selecting value vall from object O, which ex
 delayed-init-ref.scala:19: warning: Selecting value vall from object O, which extends scala.DelayedInit, is likely to yield an uninitialized value
   println(vall)       // warn
           ^
+delayed-init-ref.scala:28: warning: trait DelayedInit in package scala is deprecated: DelayedInit semantics can be surprising.
+(For details and a proposed alternative, see https://issues.scala-lang.org/browse/SI-4330?jql=labels%20%3D%20delayedinit%20AND%20resolution%20%3D%20unresolved. Support for `App` will continue.)
```
Comment (gkossakowski) [inline test/files/neg/delayed-init-ref.check]: You forgot to update the test so it will fail.

Comment (adriaanm) [inline test/files/neg/delayed-init-ref.check]: thanks -- not a great idea to code while getting ready for the airport

**Comment by adriaanm** [comment]
check file updated :)

**Comment by gkossakowski** [comment]
LGTM

**Comment by etorreborre** [comment]
Just FYI, specs2 is using `DelayedInit` to enable this use case:

``` scala
"this url must be ok" >> new Context {
   // whatever needs to be done
   // and potentially throws Exceptions
}
```

I need `Context` to extend `DelayedInit` in order to case test failures (and possibly trigger before/after behaviour). I hope there will be an alternative to that when `DelayedInit` disappears.

**Comment by adriaanm** [comment]
Thanks for the use case. The primary goal of the deprecation is to warn
people. We won't actually remove it until we've provided an alternative
that covers usages like these.

**Pull request merged**